factory_reset/reset.proto

```diff
@@ -0,0 +1,41 @@
+// This file defines a gNOI API used for factory reseting a Target.
+syntax = "proto3";
+
+package gnoi.factory_reset;
+
+import "github.com/openconfig/gnoi/types/types.proto";
+
+option (types.gnoi_version) = "0.1.0";
+
+// The FactoryReset service exported by Targets.
+service FactoryReset {
+  // The Initiate RPC allows the Client to instruct the Target to immediately
+  // clean all existing state and boot the Target in the same condition as it is
+  // shipped from factory. State includes storage, configuration, logs,
+  // certificates and licenses.
+  //
+  // It is optional for the Target to support rolling back the OS to the same
+  // version shipped from factory. If not supported, the running OS should be
+  // used.
+  //
+  // It is optional for the Target to zero fill permanent storage where state
+  // data is stored.
+  rpc Initiate(InitiateRequest) returns (InitiateResponse);
+}
+
+message InitiateRequest {
+  // Instructs the Target to rollback the OS to the same version as it shipped
+  // from factory.
+  bool factory_os = 1;
+  // Instructs the Target to zero fill persistent storage state data.
+  bool zero_fill = 2;
+}
+
+message InitiateResponse {
+  // Must be set when factory_os is set and the Target does not support it.
+  // Does not halt the Factory Reset process.
+  bool warning_unsupported_factory_os = 1;
```

Comment:
Is it more canonical to return an error code (e.g., `INVALID_ARGUMENT`) and give this sort of information in the error message / details ?

Reply:
The reasoning is that if a client wants to factory reset a Target, neither the fact that it does not return to factory os, or perform zero fill, would be reason to not proceed with the Factory Reset. Additionally a client can blanket Reset RPC a series of different Targets without having to store intent about zero fill or factory os capabilities. It will however have that information upon every Reset RPC. I am not incredibly strong about this, it just seems to me the least complex option given the use case.

Reply:
@gcsl PTAL (and add refs to canonical error handling please)

Reply:
I suggest returning an explicit error if the device doesn't support the requested options and not proceed with a reset.  gRPC does allow returning a proto message as part of the error in the "details" field as Anees indicated, which could include the current message with booleans to indicate what wasn't accepted.

https://cloud.google.com/apis/design/errors#error_model

Reply:
I have updated the service with documentation that states the Target should return gRPC error code UNIMPLEMENTED when either factory_os or zero_fill are requested but not implemented.